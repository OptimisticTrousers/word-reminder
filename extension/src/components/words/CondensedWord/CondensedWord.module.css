.word__phonetic,
.word__button,
.word__meanings,
.word__subtitle,
.word__block,
.word__line,
.word__definition {
  margin-top: 0.75rem;
}

.word__button {
  display: flex;
  align-items: center;
}

.word__item,
.word__definition,
.word__phonetic {
  list-style: none;
}

.word__block,
.word__definition {
  padding: 1rem;
  border-radius: 5px;
  border: 1px solid var(--border-color);
  background-color: var(--primary-card-bkg-color);
}

.word__text {
  margin-right: 0.75rem;
}
