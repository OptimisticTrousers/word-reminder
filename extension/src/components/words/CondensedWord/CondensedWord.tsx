import { Definition, Detail, Meaning, Phonetic } from "common";
import CSSModules from "react-css-modules";
import { Play } from "lucide-react";

import styles from "./CondensedWord.module.css";

export const CondensedWord = CSSModules(
  function ({ details }: { details: Detail[] }) {
    function handleAudio(url: string) {
      const audio = new Audio(url);
      audio.play();
    }

    return (
      <div styleName="word">
        {details.map((detail: Detail) => {
          return (
            <>
              <div styleName="word__top">
                <div styleName="word__title">
                  {detail.phonetic && (
                    <p styleName="word__phonetic">{detail.phonetic}</p>
                  )}
                  <ul styleName="word__phonetics">
                    {detail.phonetics &&
                      detail.phonetics.map((phonetic: Phonetic) => {
                        return (
                          <li styleName="word__phonetic">
                            {phonetic.text && (
                              <p styleName="word__text">{phonetic.text}</p>
                            )}
                            {phonetic.audio && (
                              <button
                                styleName="word__button"
                                onClick={() => handleAudio(phonetic.audio!)}
                              >
                                <span styleName="word__text">
                                  Pronounce word
                                </span>
                                <Play styleName="word__icon" />
                              </button>
                            )}
                          </li>
                        );
                      })}
                  </ul>
                </div>
              </div>
              <div styleName="word__container">
                {detail.meanings.map((meaning: Meaning, index: number) => (
                  <div styleName="word__meanings" key={index}>
                    <span styleName="word__part-of-speech">
                      {meaning.partOfSpeech} <hr styleName="word__line" />
                    </span>
                    <div styleName="word__block">
                      <h3 styleName="word__subtitle">Meanings</h3>
                      <ul styleName="word__definitions">
                        {meaning.definitions.map(
                          (definition: Definition, index: number) => (
                            <li styleName="word__definition" key={index}>
                              <p styleName="word__definition">
                                {definition.definition}
                              </p>
                              {definition.example && (
                                <p styleName="word__example">
                                  {definition.example}
                                </p>
                              )}
                            </li>
                          )
                        )}
                      </ul>
                    </div>
                  </div>
                ))}
              </div>
            </>
          );
        })}
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
