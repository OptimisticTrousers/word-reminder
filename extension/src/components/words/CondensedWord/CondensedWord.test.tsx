import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { CondensedWord } from "./CondensedWord";

describe("CondensedWord", () => {
  const details = [
    {
      word: "word",
      phonetic: "/wɜːd/",
      phonetics: [
        {
          text: "/wɜːd/",
          audio: "",
        },
        {
          text: "/wɝd/",
          audio:
            "https://api.dictionaryapi.dev/media/pronunciations/en/word-us.mp3",
        },
      ],
      meanings: [
        {
          partOfSpeech: "noun",
          definitions: [
            {
              definition: "News; tidings (used without an article).",
              example: "Have you had any word from John yet?",
            },
            {
              definition:
                "An order; a request or instruction; an expression of will.",
              synonyms: ["promise"],
              antonyms: ["false promise"],
              example: "Don't fire till I give the word",
            },
          ],
          synonyms: ["Bible", "Logos", "vocable"],
          antonyms: [],
        },
        {
          partOfSpeech: "verb",
          definitions: [
            {
              definition:
                "To say or write (something) using particular words; to phrase (something).",
              example: "I’m not sure how to word this letter to the council.",
            },
            {
              definition: "To flatter with words, to cajole.",
            },
          ],
        },
        {
          partOfSpeech: "interjection",
          definitions: [
            {
              definition:
                "(stereotypically) An abbreviated form of word up; a statement of the acknowledgment of fact with a hint of nonchalant approval.",
            },
          ],
        },
      ],
    },
    {
      word: "word",
      phonetic: "/wɜːd/",
      phonetics: [
        {
          text: "/wɜːd/",
          audio: "",
        },
        {
          text: "/wɝd/",
          audio:
            "https://api.dictionaryapi.dev/media/pronunciations/en/word-us.mp3",
          sourceUrl: "https://commons.wikimedia.org/w/index.php?curid=1197728",
          license: {
            name: "BY-SA 3.0",
            url: "https://creativecommons.org/licenses/by-sa/3.0",
          },
        },
      ],
      meanings: [
        {
          partOfSpeech: "verb",
          definitions: [
            {
              definition: "(except in set phrases) To be, become, betide.",
              example: "Well worth thee, me friend.",
            },
          ],
        },
      ],
    },
  ];

  it("renders meanings with part of speech, definitions, examples only", async () => {
    const { asFragment } = render(<CondensedWord details={details} />);

    const phonetic = screen.getAllByText("/wɜːd/");
    const nounPartOfSpeech = screen.getByText("noun");
    const definition1 = screen.getByText(
      "News; tidings (used without an article)."
    );
    const example1 = screen.getByText("Have you had any word from John yet?");
    const definition2 = screen.getByText(
      "An order; a request or instruction; an expression of will."
    );
    const example2 = screen.getByText("Don't fire till I give the word");
    const verbPartOfSpeech = screen.getAllByText("verb");
    const definition3 = screen.getByText(
      "To say or write (something) using particular words; to phrase (something)."
    );
    const example3 = screen.getByText(
      "I’m not sure how to word this letter to the council."
    );
    const definition4 = screen.getByText("To flatter with words, to cajole.");
    const interjectionPartOfSpeech = screen.getByText("interjection");
    const definition5 = screen.getByText(
      "(stereotypically) An abbreviated form of word up; a statement of the acknowledgment of fact with a hint of nonchalant approval."
    );
    const definition6 = screen.getByText(
      "(except in set phrases) To be, become, betide."
    );
    const word = screen.queryByRole("heading", { name: "word" });
    const link = screen.queryByRole("link");
    const example4 = screen.queryByText("Well worth thee, me friend.");
    const synonym1 = screen.queryByText("promise");
    const synonym2 = screen.queryByText("Bible");
    const synonym3 = screen.queryByText("Logos");
    const synonym4 = screen.queryByText("vocable");
    const antonym1 = screen.queryByText("false promise");

    expect(word).not.toBeInTheDocument();
    expect(link).not.toBeInTheDocument();
    expect(synonym1).not.toBeInTheDocument();
    expect(synonym2).not.toBeInTheDocument();
    expect(synonym3).not.toBeInTheDocument();
    expect(synonym4).not.toBeInTheDocument();
    expect(antonym1).not.toBeInTheDocument();
    expect(phonetic).toHaveLength(4);
    expect(nounPartOfSpeech).toBeInTheDocument();
    expect(definition1).toBeInTheDocument();
    expect(example1).toBeInTheDocument();
    expect(definition2).toBeInTheDocument();
    expect(example2).toBeInTheDocument();
    expect(verbPartOfSpeech).toHaveLength(2);
    expect(definition3).toBeInTheDocument();
    expect(example3).toBeInTheDocument();
    expect(definition4).toBeInTheDocument();
    expect(interjectionPartOfSpeech).toBeInTheDocument();
    expect(definition5).toBeInTheDocument();
    expect(definition6).toBeInTheDocument();
    expect(example4).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("plays audio to pronounce word when available", async () => {
    const details = [
      {
        word: "word",
        phonetic: "/wɜːd/",
        phonetics: [
          {
            text: "/wɜːd/",
            audio: "",
          },
          {
            text: "/wɝd/",
            audio:
              "https://api.dictionaryapi.dev/media/pronunciations/en/word-us.mp3",
          },
        ],
        meanings: [
          {
            partOfSpeech: "noun",
            definitions: [
              {
                definition: "News; tidings (used without an article).",
                example: "Have you had any word from John yet?",
              },
            ],
          },
        ],
      },
      {
        word: "word",
        phonetic: "/wɜːd/",
        phonetics: [
          {
            text: "/wɜːd/",
            audio: "",
          },
          {
            text: "/wɝd/",
            audio:
              "https://api.dictionaryapi.dev/media/pronunciations/en/word-us.mp3",
          },
        ],
        meanings: [
          {
            partOfSpeech: "verb",
            definitions: [
              {
                definition: "(except in set phrases) To be, become, betide.",
                example: "Well worth thee, me friend.",
              },
            ],
          },
        ],
      },
    ];
    const mockPlay = vi.fn();
    const mockCtor = vi.fn();
    globalThis.Audio = mockCtor.mockImplementation(() => {
      return {
        play: mockPlay,
      };
    });
    const user = userEvent.setup();

    render(<CondensedWord details={details} />);
    const audioButton = screen.getAllByRole("button", {
      name: "Pronounce word",
    })[0];
    await user.click(audioButton);

    expect(mockCtor).toHaveBeenCalledTimes(1);
    expect(mockCtor).toHaveBeenCalledWith(
      "https://api.dictionaryapi.dev/media/pronunciations/en/word-us.mp3"
    );
    expect(mockPlay).toHaveBeenCalledTimes(1);
    expect(mockPlay).toHaveBeenCalledWith();
  });
});
