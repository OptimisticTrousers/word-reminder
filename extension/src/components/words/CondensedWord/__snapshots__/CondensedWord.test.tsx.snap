// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CondensedWord > renders meanings with part of speech, definitions, examples only 1`] = `
<DocumentFragment>
  <div
    class="_words_09a47a"
  >
    <div
      class="_word_09a47a"
    >
      <div
        class="_word__top_09a47a"
      >
        <div
          class="_word__title_09a47a"
        >
          <p
            class="_word__phonetic_09a47a"
          >
            /wɜːd/
          </p>
          <ul
            class="_word__phonetics_09a47a"
          >
            <li
              class="_word__phonetic_09a47a"
            >
              <p
                class="_word__text_09a47a"
              >
                /wɜːd/
              </p>
            </li>
            <li
              class="_word__phonetic_09a47a"
            >
              <p
                class="_word__text_09a47a"
              >
                /wɝd/
              </p>
              <button
                class="_word__button_09a47a"
              >
                <span
                  class="_word__text_09a47a"
                >
                  Pronounce word
                </span>
                <svg
                  class="lucide lucide-play _word__icon_09a47a"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="6 3 20 12 6 21 6 3"
                  />
                </svg>
              </button>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="_word__container_09a47a"
      >
        <div
          class="_word__meanings_09a47a"
        >
          <span
            class="_word__part-of-speech_09a47a"
          >
            noun 
            <hr
              class="_word__line_09a47a"
            />
          </span>
          <div
            class="_word__block_09a47a"
          >
            <h3
              class="_word__subtitle_09a47a"
            >
              Meanings
            </h3>
            <ul
              class="_word__definitions_09a47a"
            >
              <li
                class="_word__definition_09a47a"
              >
                <p
                  class="_word__definition_09a47a"
                >
                  News; tidings (used without an article).
                </p>
                <p
                  class="_word__example_09a47a"
                >
                  Have you had any word from John yet?
                </p>
              </li>
              <li
                class="_word__definition_09a47a"
              >
                <p
                  class="_word__definition_09a47a"
                >
                  An order; a request or instruction; an expression of will.
                </p>
                <p
                  class="_word__example_09a47a"
                >
                  Don't fire till I give the word
                </p>
              </li>
            </ul>
          </div>
        </div>
        <div
          class="_word__meanings_09a47a"
        >
          <span
            class="_word__part-of-speech_09a47a"
          >
            verb 
            <hr
              class="_word__line_09a47a"
            />
          </span>
          <div
            class="_word__block_09a47a"
          >
            <h3
              class="_word__subtitle_09a47a"
            >
              Meanings
            </h3>
            <ul
              class="_word__definitions_09a47a"
            >
              <li
                class="_word__definition_09a47a"
              >
                <p
                  class="_word__definition_09a47a"
                >
                  To say or write (something) using particular words; to phrase (something).
                </p>
                <p
                  class="_word__example_09a47a"
                >
                  I’m not sure how to word this letter to the council.
                </p>
              </li>
              <li
                class="_word__definition_09a47a"
              >
                <p
                  class="_word__definition_09a47a"
                >
                  To flatter with words, to cajole.
                </p>
              </li>
            </ul>
          </div>
        </div>
        <div
          class="_word__meanings_09a47a"
        >
          <span
            class="_word__part-of-speech_09a47a"
          >
            interjection 
            <hr
              class="_word__line_09a47a"
            />
          </span>
          <div
            class="_word__block_09a47a"
          >
            <h3
              class="_word__subtitle_09a47a"
            >
              Meanings
            </h3>
            <ul
              class="_word__definitions_09a47a"
            >
              <li
                class="_word__definition_09a47a"
              >
                <p
                  class="_word__definition_09a47a"
                >
                  (stereotypically) An abbreviated form of word up; a statement of the acknowledgment of fact with a hint of nonchalant approval.
                </p>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
    <div
      class="_word_09a47a"
    >
      <div
        class="_word__top_09a47a"
      >
        <div
          class="_word__title_09a47a"
        >
          <p
            class="_word__phonetic_09a47a"
          >
            /wɜːd/
          </p>
          <ul
            class="_word__phonetics_09a47a"
          >
            <li
              class="_word__phonetic_09a47a"
            >
              <p
                class="_word__text_09a47a"
              >
                /wɜːd/
              </p>
            </li>
            <li
              class="_word__phonetic_09a47a"
            >
              <p
                class="_word__text_09a47a"
              >
                /wɝd/
              </p>
              <button
                class="_word__button_09a47a"
              >
                <span
                  class="_word__text_09a47a"
                >
                  Pronounce word
                </span>
                <svg
                  class="lucide lucide-play _word__icon_09a47a"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="6 3 20 12 6 21 6 3"
                  />
                </svg>
              </button>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="_word__container_09a47a"
      >
        <div
          class="_word__meanings_09a47a"
        >
          <span
            class="_word__part-of-speech_09a47a"
          >
            verb 
            <hr
              class="_word__line_09a47a"
            />
          </span>
          <div
            class="_word__block_09a47a"
          >
            <h3
              class="_word__subtitle_09a47a"
            >
              Meanings
            </h3>
            <ul
              class="_word__definitions_09a47a"
            >
              <li
                class="_word__definition_09a47a"
              >
                <p
                  class="_word__definition_09a47a"
                >
                  (except in set phrases) To be, become, betide.
                </p>
                <p
                  class="_word__example_09a47a"
                >
                  Well worth thee, me friend.
                </p>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
